### PR TITLE
Ztp feature dev

### PR DESCRIPTION
**Pull request by arunsundark**
The pull request contains the backend  code and models for ztp feature implementation 

ztp test results

### CODE REVIEW

File: models/yang/openconfig-ztp.yang [hunk 1]
```diff
@@ -0,0 +1,86 @@
+module openconfig-ztp {
```
Comment (nirenjan) [inline models/yang/openconfig-ztp.yang:1]: Are we allowed to use the `openconfig` prefix for our custom Yang models? I thought we should be using separate SONiC Yang models

File: scripts/host_modules/ztp_handler.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
+#!/usr/bin/env python3
+'''
+Copyright 2019 Broadcom. The term "Broadcom" refers to Broadcom Inc.
```
Comment (nirenjan) [inline scripts/host_modules/ztp_handler.py:3]: Copyright assigned to Broadcom? Shouldn't we own the copyright for our code?

File: scripts/host_modules/ztp_handler.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+'''
+
+import sys
+import argparse
```
Comment (nirenjan) [inline scripts/host_modules/ztp_handler.py]: Do we really need to use argparse here? Can you ensure that you address any comments reported by `pylint`?

File: scripts/host_modules/ztp_handler.py [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+        
+def register():
+    """Return the class name"""
+    return ZTP
+
+
+
+ztp_cfg = None
+## Signal handler is called on SIGTERM or SIGINT
+def signal_handler(signum, frame):
```
Comment (nirenjan) [inline scripts/host_modules/ztp_handler.py]: Not sure we need this function - the top level host service should be the only entity listening on signals.

File: scripts/host_modules/ztp_handler.py [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+            rc = err.returncode
+            output = ""
+
+        return rc, output
+
+    @host_service.method(host_service.bus_name(MOD_NAME), in_signature='', out_signature='')
+    def enable(self):
+        self._run_command("enable")
+
+    @host_service.method(host_service.bus_name(MOD_NAME), in_signature='', out_signature='is')
```
Comment (nirenjan) [inline scripts/host_modules/ztp_handler.py]: Change your `out_signature` to `s`, and return only the string `disabled` or `enabled`. Change this in transformer also.

File: scripts/host_modules/ztp_handler.py [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+            return 0,"disabled"
+        else:
+            return 0,"enabled"
+
+
+    @host_service.method(host_service.bus_name(MOD_NAME), in_signature='', out_signature='')
+    def disable(self):
+        self._run_command(["disable", "-y"])
+
+    @host_service.method(host_service.bus_name(MOD_NAME), in_signature='', out_signature='is')
```
Comment (nirenjan) [inline scripts/host_modules/ztp_handler.py]: Same here - you don't use the return code, so just use `s` in the `out_signature`

File: scripts/host_modules/ztp_handler.py [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+## Helper API to check if ztp systemd service is active or not.
+def ztp_active():
+    try:
+        rc = subprocess.check_call(['systemctl', 'is-active', '--quiet', 'ztp'])
+    except subprocess.CalledProcessError as e:
+        rc = e.returncode
+        pass
+    return rc
+
+## Convert seconds to hh:mm:ss format
```
Comment (nirenjan) [inline scripts/host_modules/ztp_handler.py:113]: Format `hh:mm:ss` doesn't line up with the format_str below.

File: scripts/host_modules/ztp_handler.py [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+    return time.strftime(format_str, time.gmtime(seconds))
+
+## Calculate time diff
+def timeDiff(startTimeStamp, endTimeStamp):
+    try:
+        endTime = time.strptime(endTimeStamp.strip(), "%Y-%m-%d %H:%M:%S %Z")
+        startTime = time.strptime(startTimeStamp.strip(), "%Y-%m-%d %H:%M:%S %Z")
+        time_diff = int(time.mktime(endTime) - time.mktime(startTime))
+        timeStr = formatTime(time_diff)
+    except:
```
Comment (nirenjan) [inline scripts/host_modules/ztp_handler.py:130]: Shouldn't use blanket except, unless absolutely necessary.

File: scripts/host_modules/ztp_handler.py [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+    
+    # Print individual section ZTP status
+        keys = sorted(ztpDict.keys())
+        for k in keys:
+            v = ztpDict.get(k)
+            if isinstance(v, dict):
+                configdict[k] = {}
+                configdict[k]['cfg_sectionname'] = k
+                if v.get('description') is not None:
+                    print('Description     : %s' % v.get('description'))
```
Comment (nirenjan) [inline scripts/host_modules/ztp_handler.py]: Remove the prints

File: src/translib/transformer/xfmr_ztp.go [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+	// the host server. The application is responsible for performing
+	// type assertions to get the correct data.
+	result := hostQuery("ztp." + action)
+	if result.Err != nil {
+		return output, result.Err
+	}
+	if ((action == "status") || (action == "getcfg")) {
+		// ztp.status returns an exit code and the stdout of the command
+		// We only care about the stdout (which is at [1] in the slice)
+		output, _ = result.Body[1].(string)
```
Comment (nirenjan) [inline src/translib/transformer/xfmr_ztp.go]: Use `result.Body[0].(string)`, as described in my comment above - we don't need to return the unnecessary "code"

File: src/translib/transformer/xfmr_ztp.go [hunk 11]
```diff
@@ -0,0 +1,339 @@
...
+		log.Info("Invalid attr:",attr)
+	}
+    }
+}
+
+/* Populate ztp status ygot tree */
+
+func populateStatusYgotTree(statusObj *ocbinds.OpenconfigZtp_Ztp_State, statusCache *ztpStatusCache) {
+
+    if _,present :=statusCache.ztpStatusMap[ZTP_STATUS_ACTIVITY_STRING]; present {
```
Comment (nirenjan) [inline src/translib/transformer/xfmr_ztp.go]: `if act, present := statusCache... ; present {`

Otherwise, you are just repeating yourself.

**Review by nirenjan — CHANGES_REQUESTED** [review]

File: models/yang/openconfig-ztp.yang (hunk 1 above)
Comment (arunsundark) [reply models/yang/openconfig-ztp.yang:1]: Transformer does not support Sonic yangs that do not have one-to-one mapping with Redis Db. So an "Openconfig style yang" is used

File: scripts/host_modules/ztp_handler.py (hunk 3 above)
Comment (arunsundark) [reply scripts/host_modules/ztp_handler.py]: Will remove it

File: scripts/host_modules/ztp_handler.py (hunk 2 above)
Comment (arunsundark) [reply scripts/host_modules/ztp_handler.py:3]: A part of code in ztp_handler.py file is taken from broadcom's implementation of ztp(https://github.com/rajendra-dendukuri/sonic-ztp/blob/initial_release/src/usr/bin/ztp), so the same copyright is used.

**Review by nirenjan — APPROVED** [review]

**Pull request merged**